internal/distro/fedora/distro.go

```diff
@@ -544,44 +544,40 @@ func (t *imageType) Size(size uint64) uint64 {
 	return size
 }
 
-func (t *imageType) getPackages(name string) rpmmd.PackageSet {
-	getter := t.packageSets[name]
-	if getter == nil {
-		return rpmmd.PackageSet{}
-	}
-
-	return getter(t)
-}
-
 func (t *imageType) PackageSets(bp blueprint.Blueprint, options distro.ImageOptions, repos []rpmmd.RepoConfig) map[string][]rpmmd.PackageSet {
 	// merge package sets that appear in the image type with the package sets
 	// of the same name from the distro and arch
-	mergedSets := make(map[string]rpmmd.PackageSet)
+	packageSets := make(map[string]rpmmd.PackageSet)
 
-	imageSets := t.packageSets
-
-	for name := range imageSets {
-		mergedSets[name] = t.getPackages(name)
+	for name, getter := range t.packageSets {
+		packageSets[name] = getter(t)
 	}
 
-	if _, hasPackages := imageSets[osPkgsKey]; !hasPackages {
-		// should this be possible??
-		mergedSets[osPkgsKey] = rpmmd.PackageSet{}
+	// depsolve bp packages separately
+	// bp packages aren't restricted by exclude lists
+	// do not include the kernel package as it is included from the pipeline
+	packageSets[blueprintPkgsKey] = rpmmd.PackageSet{Include: bp.GetPackagesEx(false)}
+
+	// amend with repository information
+	for name := range packageSets {
+		packageSet := packageSets[name]
+		for _, repo := range repos {
+			if repo.AppliesToPackageSet(name) {
+				packageSet.Repositories = append(packageSet.Repositories, repo)
+			}
+		}
+		packageSets[name] = packageSet
```

Comment:
The triple loop here (which includes the inner `AppliestoPackageSet()` loop) seems a bit hard to follow and maybe unnecessarily loopy?

I think this would be equivalent and doesn't need the repo helper:
```go
	for _, repo := range repos {
		for _, psName := range repo.PackageSets {
			ps := packageSets[psName]
			ps.Repositories = append(ps.Repositories, repo)
			packageSets[psName] = ps
		}
	}
```

Reply:
D'oh, you are so right. I'll fix this (it also fixes a bug in a corner case).

Reply:
Fixed.